dropshot_endpoint/src/metadata.rs

```diff
@@ -314,9 +328,72 @@ impl ValidatedEndpointMetadata {
 #[derive(Debug)]
 pub(crate) enum VersionRange {
     All,
-    From(semver::Version),
-    Until(semver::Version),
-    FromUntil(semver::Version, semver::Version),
+    From(VersionSpecifier),
+    Until(VersionSpecifier),
+    FromUntil(VersionSpecifier, VersionSpecifier),
+}
+
+impl syn::parse::Parse for VersionRange {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(syn::Token![..]) {
+            let _ = input.parse::<syn::Token![..]>()?;
+            if input.is_empty() {
+                Ok(VersionRange::All)
+            } else {
+                let latest = input.parse::<VersionSpecifier>()?;
+                Ok(VersionRange::Until(latest))
+            }
+        } else {
+            let earliest = input.parse::<VersionSpecifier>()?;
+            let dotdot = input.parse::<syn::Token![..]>()?;
+            let lookahead = input.lookahead1();
+            if lookahead.peek(syn::LitStr) || lookahead.peek(syn::Ident) {
+                let latest = input.parse::<VersionSpecifier>()?;
+                // If both endpoints are literals, we can check if they're
+                // in the right order.
+                if let (
+                    VersionSpecifier::Literal(earliest_semver),
+                    VersionSpecifier::Literal(latest_semver),
+                ) = (&earliest, &latest)
+                {
+                    let span = dotdot.to_token_stream();
+                    if latest_semver < earliest_semver {
+                        return Err(syn::Error::new_spanned(
+                            span,
+                            format!(
+                                "\"from\" version ({}) must be earlier than \
+                                 \"until\" version ({})",
+                                earliest_semver, latest_semver,
+                            ),
+                        ));
+                    }
+                }
+                Ok(VersionRange::FromUntil(earliest, latest))
+            } else {
+                Ok(VersionRange::From(earliest))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum VersionSpecifier {
+    Literal(semver::Version),
+    Identifier(syn::Path),
+}
+
+impl syn::parse::Parse for VersionSpecifier {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(syn::LitStr) {
+            let s = input.parse::<syn::LitStr>()?;
+            let v = parse_semver(&s)?;
+            Ok(VersionSpecifier::Literal(v))
+        } else {
+            Ok(VersionSpecifier::Identifier(input.parse::<syn::Path>()?))
```

Comment:
Good call to have a Path rather than an Ident. In the future it would seem reasonable to have an Expr, but agreed that there's no need now